pkg/provider/loadbalancer_test.go

```diff
@@ -121,377 +125,324 @@ func cmpLoadBalancerStatuses(a, b *corev1.LoadBalancerStatus) bool {
+			Entry(fmt.Sprintf("Should return loadbalancer with %s name ", "a6fa1a50e2ad011e9b210d663bd873d9"), types.UID("6fa1a50e-2ad0-11e9-b210-d663bd873d93"), "cluster1", "a6fa1a50e2ad011e9b210d663bd873d9"),
+			Entry(fmt.Sprintf("Should return loadbalancer with %s name ", "a6fa1a6582ad011e9b210d663bd873d9"), types.UID("6fa1a658-2ad0-11e9-b210-d663bd873d93"), "cluster2", "a6fa1a6582ad011e9b210d663bd873d9"),
+		)
 
-	loadBalancerIP := "123.456.7.8"
+		AfterAll(func() {
+			ctrl.Finish()
+		})
 
-	for _, test := range tests {
-		port := 30001
-		if test.portsChanged {
-			port = 30002
-		}
-		infraServiceExist := generateInfraService(
-			tenantService,
-			[]corev1.ServicePort{
-				{Name: "port1", Protocol: corev1.ProtocolTCP, Port: 80, TargetPort: intstr.IntOrString{Type: intstr.Int, IntVal: int32(port)}},
-			},
+	})
+
+	Context("With ensuring loadbalancer", func() {
+
+		var (
+			c              *mockclient.MockClient
+			ctrl           *gomock.Controller
+			ctx            context.Context
+			lb             *loadbalancer
+			tenantService  *corev1.Service
+			nodes          []*corev1.Node
+			loadBalancerIP string
 		)
-		infraServiceExist.Status = corev1.ServiceStatus{
-			LoadBalancer: corev1.LoadBalancerStatus{
-				Ingress: []corev1.LoadBalancerIngress{
-					{
-						IP: loadBalancerIP,
+
+		BeforeAll(func() {
+			ctrl, ctx = gomock.WithContext(context.Background(), GinkgoT())
+			c = mockclient.NewMockClient(ctrl)
+			lb = &loadbalancer{
+				namespace: "test",
+				client:    c,
+				config: LoadBalancerConfig{
+					CreationPollInterval: 1,
+				},
+			}
+
+			tenantService = &corev1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "service1",
+					UID:  types.UID("f6ebf172-2bb1-11e9-b210-d663bd873d93"),
+					Annotations: map[string]string{
+						"annotation-key-1": "annotation-val-1",
 					},
 				},
-			},
-		}
-		c.EXPECT().Get(
-			ctx,
-			client.ObjectKey{Name: "af6ebf1722bb111e9b210d663bd873d9", Namespace: "test"},
-			gomock.AssignableToTypeOf(&corev1.Service{}),
-		).Do(func(ctx context.Context, key client.ObjectKey, obj client.Object) {
-			if test.checkSvcExistErr == nil {
-				copyService(infraServiceExist, obj.(*corev1.Service))
+				Spec: corev1.ServiceSpec{
+					Type: corev1.ServiceTypeLoadBalancer,
+					Ports: []corev1.ServicePort{
+						{Name: "port1", Protocol: corev1.ProtocolTCP, Port: 80, NodePort: 30001},
+					},
+					ExternalTrafficPolicy: corev1.ServiceExternalTrafficPolicyTypeCluster,
+				},
 			}
-		}).Return(test.checkSvcExistErr)
-		if test.checkSvcExistErr == nil || test.checkSvcExistErr == notFoundErr {
-			infraService1 := generateInfraService(
+			nodes = []*corev1.Node{}
+			loadBalancerIP = "123.456.7.8"
+
+		})
+
+		DescribeTable("Ensure loadbalancer", func(checkSvcExistErr error, getCount int, portsChanged bool, updateSvcErr error, createSvcErr error, pollSvcErr error, expectedError error) {
```

Comment:
This test is too complex, we should divide that into multiple scenarios instead of implementing a complex logic for `DescribeTable`. Let's cover this with `It("Should <description of the case>")`.